File: src/main/org/nlogo/prim/gui/_moviestart.java
package org.nlogo.prim.gui;

import org.nlogo.api.LogoException;
import org.nlogo.nvm.EngineException;
import org.nlogo.nvm.Syntax;
import org.nlogo.window.GUIWorkspace;

public final strictfp class _moviestart
    extends org.nlogo.nvm.Command {
  @Override
  public Syntax syntax() {
    return Syntax.commandSyntax
        (new int[]{Syntax.TYPE_STRING});
  }

  @Override
  public void perform(final org.nlogo.nvm.Context context)
      throws LogoException {
    if (!(workspace instanceof GUIWorkspace)) {
      throw new EngineException(
          context, this, token().name() + " can only be used in the GUI");
    }
    final String filePath;
    try {
      filePath =
          workspace.fileManager().attachPrefix
              (argEvalString(context, 0));
    } catch (java.net.MalformedURLException ex) {
      throw new EngineException
          (context, _moviestart.this,
              argEvalString(context, 0) + " is not a valid path name: " + ex.getMessage());
    }
    workspace.waitFor
        (new org.nlogo.api.CommandRunnable() {
          public void run() throws LogoException {

            org.nlogo.awt.MovieEncoder encoder =
                ((GUIWorkspace) workspace).movieEncoder;

            if (encoder != null) {
              throw new EngineException
                  (context, _moviestart.this,
                      "There is already a movie being made. Must call MOVIE-CLOSE or MOVIE-CANCEL");
            }
            ((GUIWorkspace) workspace).movieEncoder =
                new org.nlogo.awt.JMFMovieEncoder(15, filePath);
          }
        });
    context.ip = next;
  }
}

File: netlogo-gui/src/main/prim/gui/_moviestart.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.gui

import org.nlogo.api.Syntax
import org.nlogo.nvm.{ Command, Context, EngineException }
import org.nlogo.window.GUIWorkspace

class _moviestart extends Command {
  override def syntax =
    Syntax.commandSyntax(Array(Syntax.StringType))
  override def perform(context: Context) {
    val path = {
      val arg = argEvalString(context, 0)
      try workspace.fileManager.attachPrefix(arg)
      catch {
        case ex: java.net.MalformedURLException =>
          throw new EngineException(
            context, this, arg + " is not a valid path name: " + ex.getMessage)
      }
    }
    workspace match {
      case gw: GUIWorkspace =>
        workspace.waitFor(
          new org.nlogo.api.CommandRunnable {
            def run() {
              if (gw.movieEncoder != null)
                throw new EngineException(
                  context, _moviestart.this,
                  "There is already a movie being made. Must call MOVIE-CLOSE or MOVIE-CANCEL")
              gw.movieEncoder = new org.nlogo.awt.JMFMovieEncoder(15, path)
            }})
      case _ =>
        throw new EngineException(
          context, this, token.text + " can only be used in the GUI")
    }
    context.ip = next
  }
}
